From 3a1ab5a669e30a5ff9ef005fe0241b5534b426dc Mon Sep 17 00:00:00 2001
From: Richard Guy Briggs <rgb@redhat.com>
Date: Tue, 23 Sep 2014 03:59:08 -0400
Subject: [kernel] audit: allow user processes to log from another PID namespace

Message-id: <425679f0eb2d17561e250b3cd8782627cd7dea84.1411415193.git.rgb@redhat.com>
Patchwork-id: 96455
O-Subject: [RHEL7 PATCH 5/5] audit: allow user processes to log from another PID namespace
Bugzilla: 1010455
RH-Acked-by: Paul Moore <pmoore@redhat.com>

Still only permit the audit logging daemon and control to operate from the
initial PID namespace, but allow processes to log from another PID namespace.

Upstream: 5a3cb3b
Cc: "Eric W. Biederman" <ebiederm@xmission.com>
(informed by ebiederman's c776b5d2)
Signed-off-by: Richard Guy Briggs <rgb@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/kernel/audit.c b/kernel/audit.c
index f2d919f..f724890 100644
--- a/kernel/audit.c
+++ b/kernel/audit.c
@@ -583,7 +583,7 @@ static int audit_netlink_ok(struct sk_buff *skb, u16 msg_type)
 {
  int err = 0;
 
- /* Only support the initial namespaces for now. */
+ /* Only support initial user namespace for now. */
  /*
   * We return ECONNREFUSED because it tricks userspace into thinking
   * that audit was not configured into the kernel.  Lots of users
@@ -594,8 +594,7 @@ static int audit_netlink_ok(struct sk_buff *skb, u16 msg_type)
   * userspace will reject all logins.  This should be removed when we
   * support non init namespaces!!
   */
- if ((current_user_ns() != &init_user_ns) ||
-     (task_active_pid_ns(current) != &init_pid_ns))
+ if ((current_user_ns() != &init_user_ns))
   return -ECONNREFUSED;
 
  switch (msg_type) {
@@ -615,6 +614,11 @@ static int audit_netlink_ok(struct sk_buff *skb, u16 msg_type)
  case AUDIT_TTY_SET:
  case AUDIT_TRIM:
  case AUDIT_MAKE_EQUIV:
+  /* Only support auditd and auditctl in initial pid namespace
+   * for now. */
+  if ((task_active_pid_ns(current) != &init_pid_ns))
+   return -EPERM;
+
   if (!netlink_capable(skb, CAP_AUDIT_CONTROL))
    err = -EPERM;
   break;
-- 
1.7.1